### PR TITLE
feat(NODE-2939): add new hostname canonicalization opts

### PR DESCRIPTION
**Pull request by durran**
### Description
NODE-2939
Also syncs NODE-2443 and NODE-2563

Updates `CANONICALIZE_HOST_NAME` options to accept `true`, `false`, "none", "forward", "forwardAndReverse".

#### What is changing?

When canonicalizing the host name when using GSSAPI, the driver will now behave as follows for these values:
- `true` or "forwardAndReverse": Performs a forward DNS lookup of the host and a reverse lookup of the IP address to obtain the hostname. If the reverse lookup fails the driver falls back to a cname lookup.
- "forward": Performs a cname lookup of the host.
- `false` or "none": Does no hostname canonicalization.

When syncing the spec tests some other auth spec tests we hadn't synced yet were there, so I decided to leave them and skip the single failure.

##### Is there new documentation needed for these changes?
None

#### What is the motivation for this change?

DRIVERS-1803 / NODE-2939

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/cmap/auth/gssapi.ts
Comment (addaleax) [inline src/cmap/auth/gssapi.ts:211]: > I think this should also be

```suggestion
    if (err) return callback(undefined, host);
```

> because Node.js gives you `ENODATA` errors for domains which exist but have no CNAME records attached

(https://github.com/mongodb/node-mongodb-native/pull/3129#discussion_r799450206)

Comment (durran) [reply src/cmap/auth/gssapi.ts]: Updated to suggestion.

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/cmap/auth/mongo_credentials.ts
Comment (durran) [reply src/cmap/auth/mongo_credentials.ts:170]: We did not say to deprecate in the spec - we just changed the wording that drivers MAY support `true` and `false`.

File: src/cmap/auth/gssapi.ts
Comment (durran) [reply src/cmap/auth/gssapi.ts:217]: Removed.

File: test/unit/assorted/auth.spec.test.ts
Comment (durran) [reply test/unit/assorted/auth.spec.test.ts]: Changed to hook and added skip reason with new NODE-3986 note.

**Review by nbbeeken — DISMISSED** [review]

File: test/spec/auth/connection-string.json
Comment (dariakp) [inline test/spec/auth/connection-string.json:147]: This test and the one immediately below implement NODE-2443

File: test/spec/auth/connection-string.json
Comment (dariakp) [inline test/spec/auth/connection-string.json:256]: This is NODE-2563

File: test/spec/auth/connection-string.json
Comment (dariakp) [inline test/spec/auth/connection-string.json:441]: This is NODE-2563

File: test/unit/assorted/auth.spec.test.ts
Comment (dariakp) [inline test/unit/assorted/auth.spec.test.ts:4]: This should have been implemented as part of the closed NODE-2288 ticket but wasn't

**Review by dariakp — COMMENTED** [review]
I'm leaving some comments here because we pulled in more specs than just this ticket, so we should tag the corresponding tickets on this PR and then close them when this merges

File: src/cmap/auth/gssapi.ts
Comment (dariakp) [inline src/cmap/auth/gssapi.ts:16]: I'm wondering if we could take the same approach here as we do elsewhere with enums, (e.g., `ReadPreferenceMode`) with an `export const ... Object.freeze` and `export type ... typeof ... keyof` for consistency, that way we don't have use actual strings in our code

File: test/manual/kerberos.test.js
Comment (dariakp) [inline test/manual/kerberos.test.js:91]: For this entire set of tests, the actual test logic seems to be identical - is there more we can test for each specific test case here to make sure that the code is actually doing something different in each case?

**Review by dariakp — CHANGES_REQUESTED** [review]

File: src/cmap/auth/gssapi.ts
Comment (durran) [reply src/cmap/auth/gssapi.ts]: I've updated to use the same pattern.

```
export const CanonicalizationProperties = Object.freeze({
  on: true,
  off: false,
  none: 'none',
  forward: 'forward',
  forwardAndReverse: 'forwardAndReverse'
} as const);
```

File: test/manual/kerberos.test.js
Comment (durran) [reply test/manual/kerberos.test.js:91]: I've updated to assert on the dns spy for each of the calls.

File: test/spec/auth/connection-string.json
Comment (durran) [reply test/spec/auth/connection-string.json:147]: Linked in JIRA ticket.

File: test/spec/auth/connection-string.json
Comment (durran) [reply test/spec/auth/connection-string.json:256]: Linked in JIRA ticket.

File: test/unit/assorted/auth.spec.test.ts
Comment (durran) [reply test/unit/assorted/auth.spec.test.ts:4]: https://jira.mongodb.org/browse/NODE-3986

File: src/cmap/auth/gssapi.ts
Comment (baileympearson) [inline src/cmap/auth/gssapi.ts:232]: Is this change intentional?  This seems to change the error handling behavior to always just fallback to returning the `host` instead of propagating the error.

File: src/cmap/auth/gssapi.ts
Comment (dariakp) [inline src/cmap/auth/gssapi.ts:16]: Would keeping these as `CanonicalizationValues` instead of "properties" (or perhaps `CanonicalizationSettings`) be better? Properties could imply keys that can be set on an object, but this could just be me

File: test/manual/kerberos.test.js
Comment (dariakp) [inline test/manual/kerberos.test.js:139]: Would this and the assertions in the following case be stronger if we expect `to.not.be.called` instead of `to.not.be.calledOnce`?

File: src/cmap/auth/gssapi.ts
Comment (durran) [reply src/cmap/auth/gssapi.ts:232]: Yes this was intentional to always fallback to the provided host if the dns lookups error.

File: src/cmap/auth/gssapi.ts
Comment (dariakp) [inline src/cmap/auth/gssapi.ts:216]: Does this and the logic in L219 contradict the spec directive: "The driver MUST error if **any** lookup errors or returns no results."

File: src/cmap/auth/gssapi.ts
Comment (dariakp) [inline src/cmap/auth/gssapi.ts:224]: what's the difference between dns.lookup and dns.resolveCname?

File: test/manual/kerberos.test.js
Comment (dariakp) [reply test/manual/kerberos.test.js:91]: Could we make the test cases a bit more descriptive in terms of what the spec says should happen? (i.e., for `'none'` and `false`, `"should authenticate without performing any lookups"`, for `'forwardAndReverse'` and `true`, `"should authenticate by performing a forward and reverse lookup"`, for `'forward'`, `"should authenticate by performing a forward DNS lookup only"`). And then probably also add cases for what's expected on error if that's not too difficult to do (if sinon lets us also stub the response from the functions, we could simulate errors there). For the cases that should be identical, we could make sure the logic stays the same by doing something like 
```js
for const option of [true, 'forwardAndReverse'] { 
  context(`when the value is ${option}`, function() {
    it('should authenticate by performing a forward and reverse lookup', ... );
    it('should <error behavior> if the forward lookup fails with an error', ...);
    it('should <error behavior> if the forward lookup fails with no results', ...);
    it('should <error behavior> if the reverse lookup fails with an error', ...);
    it('should <error behavior> if the reverse lookup fails with no results', ...);
  })
}
``` 
and similarly for `[false, 'none']`, with a separate block for `'forward'` only

**Review by dariakp — CHANGES_REQUESTED** [review]

File: src/cmap/auth/gssapi.ts
Comment (durran) [reply src/cmap/auth/gssapi.ts:224]: `dns.lookup()` returns an ip address and `dns.resolveCname()` returns a host name.

**Review by dariakp — COMMENTED** [review]
Let's make sure we rebase after https://github.com/mongodb/node-mongodb-native/pull/3138 goes in, because the credential option parsing will be affected

File: src/cmap/auth/gssapi.ts
Comment (durran) [reply src/cmap/auth/gssapi.ts]: Changes to `CanonicalizationValues`

File: test/manual/kerberos.test.js
Comment (durran) [reply test/manual/kerberos.test.js]: Updated to `to.not.be.called`

**Comment by durran** [comment]
@dariakp This is now rebased and uses the test refactor suggestions.

**Comment by durran** [comment]
Investigating last 2 failures.

File: src/cmap/auth/gssapi.ts
Comment (durran) [reply src/cmap/auth/gssapi.ts:213]: Spec changes merged and verified against the way the current shell works.

**Review by dariakp — COMMENTED** [review]
@durran Thanks for updating the tests! Would it be worthwhile/possible to add the expectations on the fallback behavior at each point in each case? Just to make sure we don't accidentally break that in the future?

**Review by nbbeeken — CHANGES_REQUESTED** [review]

**Comment by durran** [comment]
> @durran Thanks for updating the tests! Would it be worthwhile/possible to add the expectations on the fallback behavior at each point in each case? Just to make sure we don't accidentally break that in the future? **Edit**: if possible, we'd also want to confirm that the final response correctly uses the reply of the respective functions (like, we'd want to stub the reply from the reverse lookup to return something different than the forward, and then make sure we use the reverse response)

We actually didn't want to stub the later case, which is why I opened https://jira.mongodb.org/browse/BUILD-14646 to get an integration test for it afterwards. For the others I'll add expectations on the spies for the fallback behaviour.

File: src/cmap/auth/gssapi.ts
Comment (durran) [reply src/cmap/auth/gssapi.ts]: Changed to `GSSAPICanonicalizationValues`.

File: src/index.ts
Comment (nbbeeken) [inline src/index.ts:179]: The enum and the type need to both be exported, we've collected the enums we export on line 87

Comment (durran) [reply src/index.ts]: Updated.

File: test/manual/kerberos.test.js
Comment (durran) [reply test/manual/kerberos.test.js]: I updated the tests now to test the further error/empty array/success cases on the reverse lookup and the cname lookup. The general case I left as-is since stubbing dns.lookup would affect the other lookup calls and not test this in isolation.

File: src/index.ts
Comment (nbbeeken) [inline src/index.ts:89]: Sorry, I had edited my comment after I posted so probably was missed, I think we should not make this plural, and we need **both** the enum export here (thanks!) and type export that was here below.

Here's an example of the code that ensuring both exports supports
```ts
import { GSSAPICanonicalizationValue } from 'mongodb';

function processKerberosOption(x: GSSAPICanonicalizationValue) { 
  // The same symbol from the one import above is both a type annotation
  // And a value that can be used as an enum
  x = GSSAPICanonicalizationValue.forwardAndReverse;
}
```

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/index.ts
Comment (durran) [reply src/index.ts]: Updated to `GSSAPICanonicalizationValue` and verified the single export works as-is.

**Review by nbbeeken — APPROVED** [review]

File: test/manual/kerberos.test.js
Comment (dariakp) [inline test/manual/kerberos.test.js]: Looks like this is still WIP - but we probably want to either uncomment these or delete them before we merge.

Also, to summarize discussions from the earlier calls, for both success and failure cases, wherever possible, it would be good to assert the result from the correct function is returned (not just counting function calls, because, among other things, function call counts don't ensure any particular order of calls) or what those functions are called with. We can do that by making sure the functions resolve with different values and then checking that the callback is called with the one from the expected function call.

**Review by dariakp — CHANGES_REQUESTED** [review]

File: test/manual/kerberos.test.js
Comment (nbbeeken) [reply test/manual/kerberos.test.js]: Just scanning I think that the 'when the cname lookup is empty' test section could test the value of the hostname returned, it would be some mock value, but the empty case at least hands back what was passed in. 

Separate point: 
Would it help to move the testing of this to a unit file and export the `performGSSAPICanonicalizeHostName` function? I think we have a good idea of the very explicit order of events that need be tested, and putting it in unit would just mean not needing what ever env variables are needed to spin up the actual int tests.

Comment (durran) [reply test/manual/kerberos.test.js]: I've uncommented that. With respect to asserting the number of function calls vs what they were called with, it's a bit tricky in this case. Firstly, these tests never stub `dns.lookup` as this is not the only place it is used and the stubs are on the entire module. So these tests are stubbing the reverse ptr lookup and the cname lookups since they are only used in the canonicalization case. With the reverse ptr lookup, that would need to assert on the stub that it was passed an IP address, which I didn't want to hard code in the test and attempting the lookup would be yet another call in the test to `dns.lookup`, which we already had a spy on. (I felt like it was going in circles). :) For the cname assertions I've changed this from `calledOnce` to `calledOnceWith` to ensure the host was being passed in.

Comment (dariakp) [reply test/manual/kerberos.test.js]: Could the stub/real method concern be worked around with the sinon callThrough method on stubs, and the conditional onFirstCall/onSecondCall type definitions for what the stub returns? (Like in the last couple of comments in this [thread](https://stackoverflow.com/questions/37343232/sinon-js-only-stub-a-method-once))

There's also Neal's suggestion to make it more of a unit test. If we have to, I don't think we absolutely must hold up the feature on this, but at the same time I am hesitant to have a feature where we don't actually fully test that the intended behavior is executed for each of the specified options.

Comment (durran) [reply test/manual/kerberos.test.js]: My preference would be to create a new ticket as a follow up for testing, and separate these into unit tests as @nbbeeken suggested by exporting the `performGSSAPICanonicalizeHostName` function. Then when the BUILD ticket is done for the DNS changes I can update the auth spec with prose tests for Kerberos and ensure our manual tests follow the prose tests.

**Review by nbbeeken — APPROVED** [review]
Thanks! LGTM

**Review by dariakp — APPROVED** [review]
LGTM - great job with the new tests!

**Pull request merged**
